drivers/can/can_xlnx.c

```diff
@@ -0,0 +1,1329 @@
+ * if such pointers are supplied by the caller. If this is the case,
+ * this function will immediately return after the transmission has
+ * been triggered, and the specified callback will be called from with-
+ * in the interrupt context once the corresponding message has been
+ * transmitted successfully. If the callback parameter is NULL, this
+ * function will block until the transmission has been confirmed, which
+ * is achieved by the interrupt context giving to a semaphore this
+ * function is trying to take from without timeout.
+ *
+ * @param dev Pointer to the CAN controller's device.
+ * @param frame Pointer to the CAN message to be transmitted.
+ * @param timeout Timeout for TX completion notification íf no callback
+ *                function pointer is provided by the caller.
+ * @param callback Callback function pointer, to be called once the
+ *                 current message has been transmitted.
+ * @param user_data Parameter data pointer to be supplied to the call-
+ *                  back function.
+ *
+ * @retval -EINVAL in case of invalid parameters. -ENETDOWN in case the
+ *         controller's current bus state is bus off. -EIO if the con-
+ *         troller's operational mode doesn't allow message transmission.
+ *         -ENOSPC if the TX FIFO is currently full. -EAGAIN in case of
+ *         TX confirmation timeout.
+ */
+int can_xlnx_send(const struct device *dev,
+		  const struct can_frame *frame,
+		  k_timeout_t timeout,
+		  can_tx_callback_t callback,
+		  void *user_data)
+{
+	uintptr_t base_addr = DEVICE_MMIO_GET(dev);
+	struct can_xlnx_dev_data *dev_data = dev->data;
+	struct can_xlnx_tx_done_handler done_handler;
+	uint32_t sr;
+	uint32_t isr;
+	uint32_t ier;
+	uint32_t idr;
+	uint32_t dlcr;
+	uint32_t dw1;
+	uint32_t dw2;
+	int sem_rc;
+
+	__ASSERT(frame != NULL, "%s %s: frame pointer is NULL",
+		 __func__, dev->name);
+
+	/* CAN FD is unsupported */
+	if (frame->fd != 0) {
+		LOG_ERR("CAN FD is unsuppoted by the Xilinx CAN controller");
+		return -EINVAL;
+	}
+
+	/* Maximum DLC value is 8. */
+	if (frame->dlc > CAN_MAX_DLC) {
+		LOG_ERR("Cannot send CAN frame via interface %s: DLC %u exceeds "
+			"maximum valid value 8", dev->name, frame->dlc);
+		return -EINVAL;
+	}
+
+	/* Cannot send if the controller is in bus off state */
+	if (dev_data->bus_off == 1) {
+		return -ENETDOWN;
+	}
+
+	/* Check if the current operational mode is suitable for transmission */
+	sr = sys_read32(base_addr + CAN_XLNX_SR_OFFSET);
+	if (sr & (CAN_XLNX_SR_CONFIG_MODE | CAN_XLNX_SR_SNOOP_MODE)) {
+		LOG_ERR("Cannot send CAN frame via interface %s: unable to transmit "
+			"in %s mode", dev->name, (sr & CAN_XLNX_SR_CONFIG_MODE) ?
+			"config" : "snoop");
+		return -EIO;
+	}
+
+	/* Cannot send more frames if the TX FIFO is full */
+	isr = sys_read32(base_addr + CAN_XLNX_ISR_OFFSET);
+	if (isr & CAN_XLNX_IRQ_TX_FULL) {
+		LOG_ERR("Cannot send CAN frame via interface %s: TX FIFO full", dev->name);
+		return -ENOSPC;
+	}
```

Comment:
You need to honor the timeout for acquiring a free TX mailbox/FIFO slot here instead of just failing. But since the driver needs to be migrated to only using one TX FIFO slot anyways, this needs to rewritten.